OpenRA.Mods.Common/Graphics/BorderedRegion.cs

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Primitives;
+
+namespace OpenRA.Mods.Common.Graphics
+{
+	public readonly struct BorderedRegion : IRenderable, IFinalizedRenderable
+	{
+		// Maps a cell offset to the index of the corner (in the 'Corner' arrays in the MapGrid.CellRamp structs)
+		// from which a border should be drawn. The index of the end corner will be (cornerIndex + 1) % 4.
+		static readonly Dictionary<CVec, int> Offset2CornerIndex = new()
+		{
+			{ new CVec(0, -1), 0 },
+			{ new CVec(1,  0), 1 },
+			{ new CVec(0,  1), 2 },
+			{ new CVec(-1, 0), 3 },
+		};
+
+		readonly CPos[] region;
+		readonly Color color, contrastColor;
+		readonly float width, contrastWidth;
+
+		public BorderedRegion(CPos[] region, Color color, float width, Color contrastColor, float contrastWidth)
+		{
+			this.region = region;
+			this.color = color;
+			this.contrastColor = contrastColor;
+			this.width = width;
+			this.contrastWidth = contrastWidth;
+		}
+
+		readonly WPos IRenderable.Pos { get { return WPos.Zero; } }
+		readonly int IRenderable.ZOffset { get { return 0; } }
+		readonly bool IRenderable.IsDecoration { get { return true; } }
+
+		IRenderable IRenderable.WithZOffset(int newOffset) { return new BorderedRegion(region, color, width, contrastColor, contrastWidth); }
+		IRenderable IRenderable.OffsetBy(in WVec offset) { return new BorderedRegion(region, color, width, contrastColor, contrastWidth); }
+		IRenderable IRenderable.AsDecoration() { return this; }
+
+		IFinalizedRenderable IRenderable.PrepareRender(WorldRenderer wr) { return this; }
+		void IFinalizedRenderable.Render(WorldRenderer wr) { DrawBorderedRegion(wr, region, color, width, contrastColor, contrastWidth); }
+		void IFinalizedRenderable.RenderDebugGeometry(WorldRenderer wr) { }
+		Rectangle IFinalizedRenderable.ScreenBounds(WorldRenderer wr) { return Rectangle.Empty; }
+
+		public static void DrawBorderedRegion(WorldRenderer wr, CPos[] region, Color color, float width, Color constrastColor, float constrastWidth)
+		{
+			var map = wr.World.Map;
+			var cr = Game.Renderer.RgbaColorRenderer;
+
+			foreach (var c in region)
+			{
+				var mpos = c.ToMPos(map);
+				if (!map.Height.Contains(mpos) || wr.World.ShroudObscures(c))
+					continue;
+
+				var tile = map.Tiles[mpos];
+				var ti = map.Rules.TerrainInfo.GetTerrainInfo(tile);
+				var ramp = ti?.RampType ?? 0;
+
+				var corners = map.Grid.Ramps[ramp].Corners;
+				var pos = map.CenterOfCell(c) - new WVec(0, 0, map.Grid.Ramps[ramp].CenterHeightOffset);
+
+				foreach (var o in Offset2CornerIndex)
```

Comment:
> it would be no option to calculate the border once during construction of a bordered region and be able to render that as a multi line?

I gave that a try, but there's a bunch of problems with that. 1) the shroud check requires that the set of drawn lines gets determined every time they get drawn, 2) you'd have to sort the lines so that the ending vertex of a preceding line matches the starting vertex of the following line. I haven't found a nice way to do this, and it probably wouldn't be any cheaper.

Also, in the end the color renderer just draws a bunch of quads individually anyway, so the gain of drawing a multi-line is probably negligible (I guess, no idea really what's costly in terms of drawing).

> or would the region change? i would imagine celltriggers typically remain at one position of the map?

Typically not. I suppose moving actors would normally use a ProximityTrigger. When the actor gets removed and added back to the world, the region could theoretically change, though.